tutorials/peft-curation-with-sdg/main.py

```diff
@@ -0,0 +1,390 @@
+                FilterLowScores(score_threshold=0),
+                text_field="answer_score",
+                score_type=bool,
+            ),
+        ]
+    )
+
+    # Run the CPU curation steps.
+    dataset = cpu_curation_steps(dataset)
+
+    # Define and run the GPU curation steps.
+    if args.device == "gpu":
+        # Create a text field comprised of the title, question, and answer.
+        # This field is used for finding semantically similar records and deduping them.
+        dataset.df["text"] = (
+            dataset.df["title"]
+            + "\n"
+            + dataset.df["question"]
+            + "\n"
+            + dataset.df["answer"]
+        )
+        dataset.df = dataset.df.to_backend("cudf")
+        gpu_curation_steps = Sequential(
+            [
+                semantic_dedupe,
+            ]
+        )
+
+        dataset = gpu_curation_steps(dataset)
+        # Delete the text field as it is no longer needed.
+        del dataset.df["text"]
+        dataset.df = dataset.df.to_backend("pandas")
+
+    dataset = dataset.persist()
+    df = dataset.to_pandas()
+    orig_len = len(orig_dataset.df)
+    new_len = len(df)
+
+    return df, orig_len, new_len
+
+
+def run_pipeline(args, jsonl_fp):
+    """
+    Run the curation pipeline.
+
+    Args:
+        args: The command-line arguments.
+        jsonl_fp: The file path to the JSONL file.
+
+    Returns:
+        The file path to the final curated JSONL file.
+    """
+    # Disable synthetic data generation if no API key is provided.
+    if args.api_key is None:
+        print("No API key provided. Skipping synthetic data generation.")
+        args.synth_gen_rounds = 0
+
+    synth_gen_ratio = args.synth_gen_ratio
+    synth_gen_rounds = args.synth_gen_rounds
+    synth_n_variants = args.synth_n_variants
+
+    assert (
+        synth_gen_ratio >= 0 and synth_gen_ratio <= 1
+    ), "The synthetic generation ratio must be between 0 and 1."
+    assert (
+        synth_gen_rounds >= 0
+    ), "The number of synthetic generation rounds must be a non-negative integer."
+    assert (
+        synth_n_variants >= 1
+    ), "The number of synthetic variants must be a positive integer."
+
+    if args.device == "gpu":
+        backend = "cudf"
+    else:
+        backend = "pandas"
+
+    out_dir_base = os.path.join(DATA_DIR, "curated")
+    jsonl_filename = os.path.basename(jsonl_fp)
+
+    # Create the synthetic data generator.
+    llm_client = AsyncOpenAIClient(
+        AsyncOpenAI(
+            base_url="https://integrate.api.nvidia.com/v1",
+            api_key=args.api_key,
+        )
+    )
+    synth_gen = SyntheticGenerator(
+        llm_client,
+        "nvidia/nemotron-4-340b-instruct",
+        {
+            "top_p": 0.7,
+            "max_tokens": 1024,
+            "seed": 1234,
+        },
+        "nvidia/nemotron-4-340b-reward",
+        n_variants=synth_n_variants,
+    )
+
+    with dask.config.set({"dataframe.backend": backend}):
```

Comment:
Do you need this? I tried running without the `with` statement and it seemed to work fine for me.

Reply:
@ayushdg had done things this way in one of my PEFT experiments where fuzzy dedupe was being used, so I'll let him comment on this.

Reply:
We'll keep it like this for now